smac_planner/src/a_star.cpp

```diff
@@ -349,6 +366,104 @@ bool AStarAlgorithm<NodeT>::isGoal(NodePtr & node)
+    node->motion_table.state_space), s(node->motion_table.state_space);
+  const NodeSE2::Coordinates & node_coords = node->pose;
+  const NodeSE2::Coordinates & goal_coords = goal->pose;
+  from[0] = node_coords.x;
+  from[1] = node_coords.y;
+  from[2] = node_coords.theta * node->motion_table.bin_size;
+  to[0] = goal_coords.x;
+  to[1] = goal_coords.y;
+  to[2] = goal_coords.theta * node->motion_table.bin_size;
+
+  float d = node->motion_table.state_space->distance(from(), to());
+  NodePtr prev(node);
+  // A move of sqrt(2) is guaranteed to be in a new cell
+  constexpr float sqrt_2 = std::sqrt(2.);
+  unsigned int num_pts = std::floor(d / sqrt_2) + 1;
+  std::vector<double> reals;
+  // Pre-allocate
+  unsigned int index = 0;
+  NodePtr next(nullptr);
+  float angle = 0.0;
+  Coordinates proposed_coordinates;
+  // Don't generate the first point because we are already there!
+  // And the last point is the goal, so ignore it too!
+  for (unsigned int i = 1; i < num_pts; i++) {
+    node->motion_table.state_space->interpolate(from(), to(), (double)i / num_pts, s());
+    reals = s.reals();
+    angle = reals[2] / node->motion_table.bin_size;
+    while (angle >= node->motion_table.num_angle_quantization_float) {
+      angle -= node->motion_table.num_angle_quantization_float;
+    }
+    while (angle < 0.0) {
+      angle += node->motion_table.num_angle_quantization_float;
+    }
+    // Turn the pose into a node, and check if it is valid
+    index = NodeSE2::getIndex(
+      static_cast<unsigned int>(reals[0]),
+      static_cast<unsigned int>(reals[1]),
+      static_cast<unsigned int>(angle));
+    // Get the node from the graph
+    if (node_getter(index, next)) {
+      Coordinates initial_node_coords = next->pose;
+      proposed_coordinates = {static_cast<float>(reals[0]), static_cast<float>(reals[1]), angle};
+      next->setPose(proposed_coordinates);
+      if (next->isNodeValid(_traverse_unknown) && next != prev) {
+        // Save the node, and its previous coordinates in case we need to abort
+        possible_nodes.emplace_back(next, initial_node_coords);
+        prev = next;
+      } else {
+        next->setPose(initial_node_coords);
+        for (const auto & node_pose : possible_nodes) {
+          const auto & n = node_pose.first;
+          n->setPose(node_pose.second);
+        }
+        return NodePtr(nullptr);
+      }
+    } else {
+      // Abort
+      for (const auto & node_pose : possible_nodes) {
+        const auto & n = node_pose.first;
+        n->setPose(node_pose.second);
+      }
+      return NodePtr(nullptr);
+    }
+  }
+  if (num_pts > 2) {
+    // Legitimate path - set the parent relationships - poses already set
+    NodePtr prev = node;
+    for (const auto & node_pose : possible_nodes) {
+      const auto & n = node_pose.first;
+      if (n != prev) {
+        n->parent = prev;
+      }
+      prev = n;
+    }
+    if (goal != prev) {
```

Comment:
Wouldn't this always be true since we only iterate from 1 to N-1?

Reply:
So it turns out that if you actually do the maths correctly you don't need to put in hacks like this...

The way the number of points/intervals was being calculated was incorrect. With it fixed, there is no need for these checks.